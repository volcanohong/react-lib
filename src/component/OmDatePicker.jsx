import 'date-fns';
import React, { useState } from 'react';
import Grid from '@material-ui/core/Grid';
import DateFnsUtils from '@date-io/date-fns';
import PropTypes from 'prop-types';
import {
    MuiPickersUtilsProvider,
    TimePicker,
    DatePicker,
} from '@material-ui/pickers';
import { makeStyles } from '@material-ui/core/styles';
import { Icon } from "@material-ui/core";
import Button from '@material-ui/core/Button';

const useStyles = makeStyles({
    inputContainer: {
        width: 80
    },
    iconButton: {
        width: 36,
        height: 36,
        padding: 0,
        minWidth: 36,
        marginRight: 8
    },
    icon: {
        fontSize: '2.375em'
    }
});

function OmDatePicker(props) {

    const classes = useStyles();
    const [selectedDate, setSelectedDate] = useState(!!props.date ? props.date : null);
    const [showDialog, setShowDialog] = useState(false);

    const handleDateChange = date => {
        setSelectedDate(date);
        props.onDateChange(date);
    };

    const toggleDialog = state => {
        if (!props.readOnly) {
            setShowDialog(state);
        }
    }

    return (
        <div className={props.className}>
            <MuiPickersUtilsProvider utils={DateFnsUtils}>
                <Grid container>
                    {
                        props.variant === 'date' ?
                            <Grid item className='om-flex om-flex-row om-items-center'>
                                {
                                    !!props.prefixicon ?
                                        <Button variant="contained" color="primary"
                                            onClick={() => toggleDialog(true)}
                                            classes={{ root: classes.iconButton }}>
                                            <Icon classes={{ root: classes.icon }}>today</Icon>
                                        </Button> : null
                                }
                                <DatePicker
                                    margin="normal"
                                    id={props.id}
                                    format={props.dateformat}
                                    value={selectedDate}
                                    onChange={(date) => handleDateChange(date)}
                                    open={showDialog}
                                    disabled={props.readOnly}
                                    onAccept={() => toggleDialog(false)}
                                    onClose={() => toggleDialog(false)}
                                    placeholder={props.placeholder}
                                    InputProps={{
                                        className: classes.inputContainer,
                                        onClick: () => toggleDialog(true)
                                    }}
                                />
                            </Grid>
                            : null
                    }
                    {
                        props.variant === 'time' ?
                            <Grid item className='om-flex om-flex-row om-items-center'>
                                {
                                    !!props.prefixicon ?
                                        <Button variant="contained" color="primary"
                                            onClick={() => toggleDialog(true)}
                                            classes={{ root: classes.iconButton }}>
                                            <Icon classes={{ root: classes.icon }}>schedule</Icon>
                                        </Button> : null
                                }
                                <TimePicker
                                    margin="normal"
                                    id={props.id}
                                    value={selectedDate}
                                    format={props.timeformat}
                                    onChange={(date) => handleDateChange(date)}
                                    open={showDialog}
                                    disabled={props.readOnly}
                                    onAccept={() => toggleDialog(false)}
                                    onClose={() => toggleDialog(false)}
                                    placeholder={props.placeholder}
                                    InputProps={{
                                        className: classes.inputContainer,
                                        onClick: () => toggleDialog(true)
                                    }}
                                />
                            </Grid>
                            : null
                    }

                </Grid>
            </MuiPickersUtilsProvider>
        </div>

    );
}

OmDatePicker.propTypes = {
    id: PropTypes.string,
    date: PropTypes.object,
    variant: PropTypes.string,
    prefixicon: PropTypes.bool,
    placeholder: PropTypes.string,
    dateformat: PropTypes.string,
    timeformat: PropTypes.string,
    className: PropTypes.string,
    onDateChange: PropTypes.func.isRequired,
    readOnly: PropTypes.bool
}

OmDatePicker.defaultProps = {
    variant: 'date',
    prefixicon: true,
    dateformat: 'dd-MM-yyyy',
    timeformat: 'HH:mm',
    readOnly: false
}

export default OmDatePicker;
