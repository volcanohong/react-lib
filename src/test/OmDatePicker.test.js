import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import OmDatePicker from '../component/OmDatePicker';
import {
    TimePicker,
    DatePicker,
} from '@material-ui/pickers';

configure({ adapter: new Adapter() });

function handleDateChange(date) {
    //do something with date
}

test('test Datepicker render', () => {
    const tree = renderer
        .create(<OmDatePicker id={'datePicker'} variant="date"
            onDateChange={handleDateChange} placeholder={'pick a date'} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

test('test Timepicker render', () => {
    const tree = renderer
        .create(<OmDatePicker id={'timePicker'} variant="time"
            onDateChange={handleDateChange} placeholder={'pick a time'} />)
        .toJSON();
    expect(tree).toMatchSnapshot();
});

test('test Datepicker placeholder', () => {

    const wrapper = shallow(
        <OmDatePicker id={'datePicker'} variant="date"
            onDateChange={handleDateChange} placeholder={'pick a time'} />
    );

    const datepicker = wrapper.find('#datePicker');
    expect(datepicker.prop('placeholder')).toBe('pick a time');
});

test('test Datepicker format', () => {
    const expected = 'MM/dd/yyyy';
    const wrapper = shallow(
        <OmDatePicker id={'datePicker'} variant="date"
            onDateChange={handleDateChange} dateformat={expected} />
    );

    const datepicker = wrapper.find('#datePicker');
    expect(datepicker.prop('format')).toBe(expected);
});

test('test Datepicker readonly', () => {
    const expected = true;
    const wrapper = shallow(
        <OmDatePicker id={'datePicker'} variant="date"
            onDateChange={handleDateChange} readOnly={expected} />
    );

    const datepicker = wrapper.find('#datePicker');
    expect(datepicker.prop('disabled')).toBe(expected);
});

// test('test open Dialog by click input', () => {
//     const expected = new Date();
//     const wrapper = mount(
//         <OmDatePicker id={'datePicker'} variant="date"
//             onDateChange={handleDateChange} />
//     );

//     expect(wrapper.find(DatePicker).prop('open')).toBe(false);

//     wrapper.find('input').simulate('click');

//     expect(wrapper.find(DatePicker).prop('open')).toBe(true);
// });

// test('test open Dialog by click input', () => {
//     const expected = new Date();
//     const wrapper = mount(
//         <OmDatePicker id={'datePicker'} variant="date"
//             onDateChange={handleDateChange} />
//     );

//     expect(wrapper.find(DatePicker).prop('open')).toBe(false);

//     wrapper.find('button').simulate('click');

//     expect(wrapper.find(DatePicker).prop('open')).toBe(true);
// });

// test('test date change function', () => {
//     const handleDateChangeMock = jest.fn();
//     var updatedDate = null;
//     const originalDate = new Date('2020-01-01');
//     const newDate = new Date('2020-06-06');
//     const wrapper = mount(
//         <OmDatePicker id={'datePicker'} variant="date"
//             onDateChange={handleDateChangeMock} date={originalDate} />
//     );

//     const event = {
//         target: { value: newDate }
//     };

//     expect(updatedDate).toBe(null);

//     const datePicker = wrapper.find(DatePicker);

//     expect(datePicker.prop('value')).toBe(originalDate);

//     datePicker.props().onChange(event);

//     expect(handleDateChangeMock).toBeCalledWith(event);
// });
