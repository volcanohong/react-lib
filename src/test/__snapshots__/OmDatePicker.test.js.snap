// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Datepicker render 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container"
  >
    <div
      className="MuiGrid-root om-flex om-flex-row om-items-center MuiGrid-item"
    >
      <button
        className="MuiButtonBase-root MuiButton-root makeStyles-iconButton-2 MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root makeStyles-icon-3"
          >
            today
          </span>
        </span>
      </button>
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
        format="dd-MM-yyyy"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-inputContainer-1 MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="datePicker"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="pick a date"
            readOnly={true}
            required={false}
            type="text"
            value=""
          />
        </div>
        
      </div>
    </div>
  </div>
</div>
`;

exports[`test Timepicker render 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container"
  >
    <div
      className="MuiGrid-root om-flex om-flex-row om-items-center MuiGrid-item"
    >
      <button
        className="MuiButtonBase-root MuiButton-root makeStyles-iconButton-2 MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root makeStyles-icon-3"
          >
            schedule
          </span>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
        format="HH:mm"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-inputContainer-1 MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="timePicker"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="pick a time"
            readOnly={true}
            required={false}
            type="text"
            value=""
          />
        </div>
        
      </div>
    </div>
  </div>
</div>
`;
